publication/models/distribution_list.py

```diff
@@ -0,0 +1,180 @@
+        store=True)
+    partner_id = fields.Many2one(
+        comodel_name='res.partner',
+        string='Receiving Partner',
+        required=True)
+    contract_partner_id = fields.Many2one(
+        comodel_name='res.partner',
+        string='Contract Partner',
+        required=True)
+    date_start = fields.Date(
+        string='Date start',
+        default=fields.Date.today(),
+        required=True)
+    date_end = fields.Date(
+        string='Date end')
+    copies = fields.Integer(string='Number of copies', default=1)
+    name = fields.Char(
+        compute='_compute_name_address',
+        string='Name')
+    contact_address = fields.Char(
+        compute='_compute_name_address',
+        string='Receiving address')
+    contract_count = fields.Integer(
+        string="Number to distribute",
+        compute='_compute_counts')
+    assigned_count = fields.Integer(
+        string="Number already assigned",
+        compute='_compute_counts')
+    available_count = fields.Integer(
+        string="Number still available",
+        compute='_compute_counts')
+
+    @api.onchange('product_id', 'contract_partner_id')
+    def _onchange_keyfields(self):
+        """Sets the proper domain for contract_partner.
+
+        Also enforces first selecting the publication.
+        """
+        self.ensure_one()
+        if self.contract_partner_id and not self.partner_id:
+            self.partner_id = self.contract_partner_id
+        if self.contract_partner_id and not self.product_id:
+            raise ValidationError(_(
+                "You must select a publication, before selecting"
+                " the contract partner."))
+        if not self.product_id:
+            return
+        valid_partners = []
+        line_model = self.env['account.analytic.invoice.line']
+        lines = line_model.search([('product_id', '=', self.product_id.id)])
+        for line in lines:
+            if line.analytic_account_id.partner_id.id not in valid_partners:
+                valid_partners.append(line.analytic_account_id.partner_id.id)
+        if not valid_partners:
+            raise ValidationError(_(
+                "There are no active subscriptions for this publication."))
+        partner_domain = [('id', 'in', valid_partners)]
+        return {'domain': {'contract_partner_id': partner_domain}}
+
+    @api.model
+    def create(self, vals):
+        result = super(DistributionList, self).create(vals)
+        result._limit_count()
+        return result
+
+    @api.multi
+    def write(self, vals):
+        result = super(DistributionList, self).write(vals)
+        self._limit_count()
+        return result
+
+    @api.multi
+    def _limit_count(self):
+        """Limit number of copies send to amount set in contract lines.
+
+        It should be possible to make a constrains method of this function,
+        but for inexplicable reasons this does not work.
+        """
+        for this in self:
```

Comment:
this -> copy ?

Reply:
We always use this when iterating over self. In this case it would also be confusing as we are not iterating over copies but over distribution list records, that have copies as a field.